### PR TITLE
docs: generate website documents correctly for 2.0

### LINKED ISSUE #1445: frontmatter example is not working
### Subject of the issue

See example https://mdxjs.com/advanced/api#compiler, it's expected to yield `{ title: "Hello, MDX" }` while I got `undefined`.

### Your environment

*   **OS**: macOS BigSur
*   **Packages**: 
*   **Env**: 

### Steps to reproduce

1. clone `https://github.com/chenxsan/bug-mdxjs-frontmatter`
2. run `yarn`
3. run `node index.js`

### Expected behaviour

Yields `{ title: "Hello, MDX" }`.

### Actual behaviour

Got `undefined`.

### PR DESCRIPTION
**Pull request by JounQin**
1. document remark-frontmatter@3.0 correctly for 2.0, close #1445
2. enable `eslint-plugin-mdx` correctly for `md` and `mdx` files
3. fix incorrect mdx version in website header

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -33,6 +33,10 @@ rules:
   import/no-extraneous-dependencies: error
   prettier/prettier: error
   no-console: off
+  import/no-unresolved:
+    - error
+    - ignore:
+        - '@mdx-js/react'
```
Comment (ChristianMurphy) [inline .eslintrc.yml:39]: why are eslint rules, updating in a documentation change?

File: .remarkrc.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module.exports = {
   plugins: [
-    './packages/remark-mdx',
+    'remark-frontmatter',
```
Comment (ChristianMurphy) [inline .remarkrc.js:3]: won't remark-mdx still be needed to parse MDX files?

File: docs/advanced/api.mdx [hunk 3]
```diff
@@ -130,13 +130,15 @@ const visit = require('unist-util-visit')
...
+  `
 ---
 title: Hello, MDX
 ---
 
-I <3 Markdown and JSX
-`)
+I &lt;3 Markdown and JSX {/* < must be escaped for \`@mdx-js/mdx@2\` */}
+`.trimStart() // `trimStart` is required for `remark-frontmatter@3`
+)
```
Comment (ChristianMurphy) [inline docs/advanced/api.mdx:141]: trimStart isn't specifically required, just a lack of spaces before `---`
```js
const file = vfile(`---
title: Hello, MDX
```
would also work

File: docs/guides/mdx-embed.md [hunk 4]
```diff
@@ -22,7 +22,8 @@ export default props => <MDXEmbedProvider>{props.children}</MDXEmbedProvider>
 ### Usage
 
 ```mdx
-// some-mdx-file.mdx
+<!-- some-mdx-file.mdx -->
```
Comment (ChristianMurphy) [inline docs/guides/mdx-embed.md:25]: :+1:

File: docs/getting-started/create-react-app.mdx [hunk 5]
```diff
@@ -95,8 +95,5 @@ Finally, see [Support][] if you’re still running into problems.
 [cra]: https://github.com/facebook/create-react-app
 [loader]: https://github.com/mdx-js/mdx/tree/main/packages/loader
 [babel-loader]: https://webpack.js.org/loaders/babel-loader/
-
-<!-- To do: update to point to website if that’s working -->
-
-[typescript]: https://github.com/mdx-js/mdx/blob/main/docs/advanced/typescript.mdx
+[typescript]: https://mdxjs.com/advanced/typescript
```
Comment (ChristianMurphy) [inline docs/getting-started/create-react-app.mdx:98]: :+1:

File: package.json [hunk 6]
```diff
@@ -19,9 +19,9 @@
     "docs-build": "gatsby build",
     "docs-deploy": "vercel && vercel alias $(pbpaste) mdxjs.com && vercel alias $(pbpaste) www.mdxjs.com",
     "format": "yarn lint --fix",
-    "lint": "eslint . --cache --ext js,jsx -f friendly",
+    "lint": "eslint . --cache --ext md,mdx,js,jsx -f friendly",
     "vercel-build": "yarn build && yarn docs-build",
-    "postinstall": "yarn-deduplicate || exit 0",
+    "postinstall": "yarn-deduplicate --strategy fewer && node scripts/patch || exit 0",
```
Comment (ChristianMurphy) [inline package.json:24]: :thinking: why are scripts being patched in a documentation change?

File: packages/gatsby-theme-mdx/package.json [hunk 7]
```diff
@@ -17,7 +17,8 @@
   },
   "peerDependencies": {
     "react": ">=16",
-    "react-dom": ">=16"
+    "react-dom": ">=16",
+    "gatsby": "^2.0.0"
```
Comment (ChristianMurphy) [inline packages/gatsby-theme-mdx/package.json:21]: This doesn't seem to be front matter related?
And also the existing/previous behavior is how Gatsby documents how themes should be made https://www.gatsbyjs.com/tutorial/building-a-theme/

File: scripts/patch.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+const skipNavFilename =
+  'node_modules/gatsby-theme-mdx1/src/components/skip-nav.js'
+
+const skipNavStyles = 'import "@reach/skip-nav/styles.css"'
+
+const skipNavContent = fs.readFileSync(skipNavFilename, 'utf8')
+
+if (!skipNavContent.startsWith(skipNavStyles)) {
+  fs.writeFileSync(skipNavFilename, skipNavStyles + '\n' + skipNavContent)
+}
```
Comment (ChristianMurphy) [inline scripts/patch.js:28]: could you expand on why this is needed?
It seems like versions are something which should be managed by package.json?
It also doesn't seem to be related to the PR's stated goal of frontmatters?

**Review by ChristianMurphy — CHANGES_REQUESTED** [review]

File: docs/advanced/api.mdx (hunk 3 above)
Comment (JounQin) [reply docs/advanced/api.mdx:141]: That’s the meaning to use `trimStart()`.

Comment (ChristianMurphy) [reply docs/advanced/api.mdx:141]: Yes, that's my point, the comment

```js
// `trimStart` is required for `remark-frontmatter@3`
```

says that `trimStart` is required, it isn't.

Something like:

```js
// there should be no spaces before frontmatter, this can be guaranteed with `trimStart`

File: scripts/patch.js (hunk 8 above)
Comment (JounQin) [reply scripts/patch.js:28]: I’ve commented at the front.

The document is generated by gastby, and it is not compatible with mdx@2 for now, so I hacked to use `gastby-theme-mdx1@npm:gastby-theme-mdx@^1.0.0` to build docs successfully.

And it requires `@mdxjs/react@1`, which results current version in title is always 1.x even it’s for 2.0.

So I have to hack like this to generate document correctly.

File: package.json (hunk 6 above)
Comment (JounQin) [reply package.json:24]: See https://github.com/mdx-js/mdx/pull/1490#discussion_r600558337

File: .eslintrc.yml (hunk 1 above)
Comment (JounQin) [reply .eslintrc.yml:39]: Because I found it’s not working actually.

File: .remarkrc.js (hunk 2 above)
Comment (JounQin) [reply .remarkrc.js:3]: `eslint-plugin-mdx` enables `remark-mdx@1` automatically. (Because of gastby, we are still using mdx@1).

See also https://github.com/mdx-js/mdx/pull/1490#discussion_r600558337

File: packages/gatsby-theme-mdx/package.json (hunk 7 above)
Comment (JounQin) [reply packages/gatsby-theme-mdx/package.json:21]: https://www.gatsbyjs.com/tutorial/building-a-theme/#add-development-dependencies-to-gatsby-theme-events

Comment (JounQin) [reply packages/gatsby-theme-mdx/package.json:21]: It results conflict gastby versions when I tried to upgrade gastby to v3 to test if it’s compatible with remark-mdx@2 now.

File: docs/advanced/api.mdx [hunk 9]
```diff
@@ -130,13 +130,15 @@ const visit = require('unist-util-visit')
...
+const file = vfile(
+  `
 ---
 title: Hello, MDX
 ---
 
-I <3 Markdown and JSX
-`)
+I &lt;3 Markdown and JSX {/* < must be escaped for \`@mdx-js/mdx@2\` */}
+`.trimStart() // `trimStart` is required for `remark-frontmatter@3`
```
Comment (JounQin) [inline docs/advanced/api.mdx:140]:
```suggestion
`.trimStart() // there should be no spaces before frontmatter for `remark-frontmatter@3`, this can be guaranteed with `trimStart`
```

@ChristianMurphy How about this?

Comment (ChristianMurphy) [reply docs/advanced/api.mdx:140]: looks good to me :+1:

File: scripts/patch.js (hunk 8 above)
Comment (ChristianMurphy) [reply scripts/patch.js:28]: Could this be solved with yarn resolutions?
Would it make sense to have this in its own PR?
It feels like this PR now captures three distinct change sets, fixing the frontmatter docs, adjusting the eslint settings, and updates to the documentation site.
All are good and welcome.

Reviewing them all together, especially when the title and description has no mention of the latter to is a bit confusing

Comment (JounQin) [reply scripts/patch.js:28]: The goal of this PR is to generate correct document for `mdx@2`.

You can check https://mdxjs.com/ and https://mdx-git-docs-frontmatter-mdx.vercel.app/ to see the difference on page header.

I’m on iPad, so I can capture screenshot for you.

File: packages/gatsby-theme-mdx/package.json (hunk 7 above)
Comment (ChristianMurphy) [reply packages/gatsby-theme-mdx/package.json:21]: Gotcha, thanks for the clarification

It feels like this PR now captures three distinct change sets, fixing the frontmatter docs, adjusting the eslint settings, and updates to the documentation site.
All are good and welcome.

Reviewing them all together, especially when the title and description has no mention of the latter to is a bit confusing

File: scripts/patch.js (hunk 8 above)
Comment (JounQin) [reply scripts/patch.js:28]: > Could this be solved with yarn resolutions?

No, we are developing `gastby-theme-mdx@2` in this repo at the same time. And I don’t know how `resolutions` be related to this patch which changes specific files in node_modules.

Comment (ChristianMurphy) [reply scripts/patch.js:28]: > So I have to hack like this to generate document correctly

Would something like https://github.com/ds300/patch-package work here?

Comment (JounQin) [reply scripts/patch.js:28]: I’ve changed the title and description to clarify.

Comment (JounQin) [reply scripts/patch.js:28]: > Would something like https://github.com/ds300/patch-package work here?

Maybe, but I’d prefer patch as-is for such simple task. (less dev pkg is always better)

Comment (ChristianMurphy) [reply scripts/patch.js:28]: Not necessarily, those packages are included in patch-package because it can detect and warn if the package it is patching has changed, and warn the user.
The patch script as is, would include the changes regardless of it they fit/work with the version of the package they are patching, with no warning of possible conficts.

Comment (JounQin) [reply scripts/patch.js:28]: `gastby-theme-mdx@1` will never change anymore. We’ve stopped developing on v1.

Comment (JounQin) [reply scripts/patch.js:28]: @ChristianMurphy Any further discussion?

Comment (ChristianMurphy) [reply scripts/patch.js:28]: I still think that using patch-packages is a clearer approach.
I'm cautious about a hand coded solution.

Comment (JounQin) [reply scripts/patch.js:28]: An extra dev dependcy is not clearer to me, and I don't think it should block this PR.

Comment (ChristianMurphy) [reply scripts/patch.js:28]: I understand your opinion and I respectfully disagree.
I see it as a well know problem with a well known solution (`patch-package`).
I don't see writing a custom solution as a worth while tradeoff to avoid some dev dependencies.

Comment (JounQin) [reply scripts/patch.js:28]: Well, if you want `patch-package` strongly, can you raise another PR for it by yourself? It is good solution, but I don't want it for such simple task in this PR.

Comment (ChristianMurphy) [reply scripts/patch.js:28]: Another mdx maintainer can approve and merge it if they want.
I'm not comfortable merging it as is.

Comment (JounQin) [reply scripts/patch.js]: So why you don't want another PR by yourself? It's just a personal taste, I'm fine if you want it and implement it by yourself, I just don't want it in this PR.

Comment (ChristianMurphy) [reply scripts/patch.js]: > I just don't want it in this PR.

Okay :man_shrugging: 
I'm still not comfortable approving and merging as is.
For the points outlined several times through this review; this PR has an overly broad scope that changes many unrelated pieces which could/should be discussed and changed independently, and it introduces some custom build scripts to patch packages.
That said, I'm also not blocking the PR either, if another maintainer feels strongly another way, they can approve and merge if they want.

> So why you don't want another PR by yourself?

If it gets merged as-is, I likely will.

**Comment by JounQin** [comment]
The only left blocker is whether to use `patch-package`, I'm -1 for this because it's insignificant.

You can raise another PR to "fix" if you want.

Merge first for now.

**Comment by ChristianMurphy** [comment]
@JounQin unified/mdx uses lazy consensus https://github.com/unifiedjs/collective/blob/c4f6485ccb0429b34572153a177c11a4ae2ada9c/decisions.md#issues
https://openoffice.apache.org/docs/governance/lazyConsensus.html

That means that PRs can be often be merged after 48 hours or so if there are not objections.
This does have an objection, it does not meet the criteria for merging, I'm reverting this change so it can go through the approval process.
You are welcome to reopen the PR and gather approvals on it.

**Comment by JounQin** [comment]
OK, let's not waste time on such things, I'm going to try `patch-package`.

But when I tried to run `yarn patch-package gatsby-theme-mdx1`.

image

It's being held, don't know is that because of `"gatsby-theme-mdx1": "npm:gatsby-theme-mdx@^1.0.0"`.

So, this means `patch-package` is not compatible here.
